### ISSUE TITLE
C++ Education Effort

### ISSUE THREAD

**Issue by ibaned** [kokkos/kokkos#1940]
This issue tracks our planning of having the Kokkos team provide C++ training seminars for Sandia staff. For all topics, the value we bring is the perspective of people close the C++ committee an compilers in terms of:
 - what is best for compile-time performance
 - what is recommended by the standards committee
 - what is idiomatic "modern" C++

Subjects personally on my mind are:
 - an understanding of lambdas, which I think are best presented as a mechanism which generates a new class type and instantiates an object of that class type.
 - basic template metaprogramming topics, this is fairly vague but essentially the common techniques for compile-time logic available in C++03, then moving into `constexpr` etc.
 - C++11 idiomatic topics including move semantics and `auto` type

**Comment by mhoemmen** [comment]
Here are some other topics that come up among my colleagues:
  - initialization rules, like the differences between `T{}` and `T()` initialization;
  - move semantics as they relate specifically to `std::unique_ptr`; and
  - remembering that `std::vector` construction and resize fill the entries.

**Comment by dhollman** [comment]
Here's my wish list I put together a while back, for reference.  I'm sure I can think of more things, too, but this is what I wrote down.
- modern C++ object design and why it matters
- RAII
- `shared_ptr<T[]>` and `unique_ptr<T[]>` instead of `vector` when you can
- what it really means to avoid global variables
- object initialization best practices
- move semantics
- ADL basics (and why you should mostly ignore it)
- no raw memory allocations
- modern C++ polymorphism, both static and dynamic
- the importance of strong ownership semantics

**Comment by mhoemmen** [comment]
A+++ on the `*_ptr<T[]>`; we might be burning double-digit percentage time on that.

**Comment by mhoemmen** [comment]
Also, what's "modern C++ object design"?  I know "pre-modern design" when I see it but what's a good reference?

**Comment by ibaned** [comment]
I don't know what @dhollman meant exactly by that, but for me it brings two things to mind:
 - with move semantics, the Rule of Three becomes the Rule of Five, or preferrably the Rule of Zero (especially with `= default` and `= delete`).
 - how to make sure objects fit some definition of triviality (e.g. trivially copyable) or constexpr, etc.

**Comment by khpierson** [comment]
I do think this is a good idea. 
Has anyone started collecting best practices for Trilinos and/or Sierra development? Also consider that  with large amounts of legacy code, there might not be immediate benefit from re-factoring to a more modern C++ design unless it is motivated by being a stepping stone towards NGP platforms.

just my 2-cents,
-Kendall

**Comment by mhoemmen** [comment]
@khpierson I think refactoring to "more modern C++ design" can happen gradually, whenever we need to change a class for other reasons.  For example, we can replace trivial constructors and destructors with `= default`, and replace the "declare constructor private but don't implement it" C++98 idiom with `= delete`.  I don't think it pays to sweep the whole code and touch everything, but we can do it whenever we are touching the code anyway.

**Comment by khpierson** [comment]
Would clang-tidy be an option for performing these operations on the code base? Perhaps we need to write SNL specific clang-tidy checks. If we had this, then at least we could identify the places where we need to modernize.

**Comment by dhollman** [comment]
Yes to all of this!  I think clang-tidy can generate incremental patches that can be applied as pull requests on a per-file basis and tested individually. Might be worth trying for a few subprojects, anyway.

On Jan 3, 2019, at 7:42 AM, Kendall H Pierson <notifications@github.com<mailto:notifications@github.com>> wrote:


Would clang-tidy be an option for performing these operations on the code base? Perhaps we need to write SNL specific clang-tidy checks. If we had this, then at least we could identify the places where we need to modernize.

**Comment by ajpowelsnl** [comment]
@PhilMiller , @crtrott , @ibaned, @nliber -- this is a great topic.  Is it something we still want to try to do?

**Comment by ibaned** [comment]
I think this would be valuable in general at Sandia. We have folks needing to make the jump from C++98 programming into C++17 and training would be beneficial.

**Comment by ajpowelsnl** [comment]
> I think this would be valuable in general at Sandia. We have folks needing to make the jump from C++98 programming into C++17 and training would be beneficial.

I wonder if it would be interesting to have such training as a "prequel" to the Kokkos Tutorials ...?

**Comment by ajpowelsnl** [comment]
@mhoemmen - would you please email me at my work address?  I like how you think :smile:

**Comment by mhoemmen** [comment]
@ajpowelsnl Awww thanks!!!  I don't have my Sandia e-mail any more, but I'm active on the Kokkos Slack channel -- please feel free to ping me there!
